src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java

```diff
@@ -0,0 +1,344 @@
+                break;
+            case TokenTypes.SLIST:
+            case TokenTypes.LITERAL_RETURN:
+                result = child.getFirstChild().getType() != TokenTypes.SEMI
+                    && checkExpr(child.getFirstChild(), args);
+                break;
+            default:
+                result = shouldCheck(expr)
+                    && checkExpr(expr.findFirstToken(TokenTypes.EXPR), args);
+                break;
+        }
+        return result;
+    }
+
+    /**
+     * Checks if given ast should be checked.
+     * Ast should have only one child expr, and have no other tokens from
+     * {@link #TOKEN_TYPES} list.
+     *
+     * @param ast token to check
+     * @return true if there is only one child expression, false otherwise
+     */
+    private static boolean shouldCheck(DetailAST ast) {
+        return ast.getChildCount(TokenTypes.EXPR) == 1
+            && ast.getChildCount(TokenTypes.SEMI) == 1
+            && !TokenUtil.findFirstTokenByPredicate(ast,
+                expr -> TOKEN_TYPES.contains(expr.getType())).isPresent();
+    }
+
+    /**
+     * Check that creation of the object can be replaced with constructor reference.
+     *
+     * @param ast token to check
+     * @param args lambda arguments
+     * @return true if token can be replaced, false otherwise
+     */
+    private static boolean checkObjectCreation(DetailAST ast, List<String> args) {
+        final DetailAST elist = ast.findFirstToken(TokenTypes.ELIST);
+        final DetailAST arrayDecl = ast.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
+        final boolean matchesCtorArgs =
+            elist != null && !isAnonClass(ast) && hasSameArgs(args, elist);
+        final boolean matchesArrayInit =
+            arrayDecl != null && args.size() == 1 && isArgUsedInArrayInit(arrayDecl, args.get(0));
+        return matchesArrayInit || matchesCtorArgs;
+    }
+
+    /**
+     * Checks that given new object creation is anon class.
+     *
+     * @param ast token to check
+     * @return true if it is anon class, false otherwise
+     */
+    private static boolean isAnonClass(DetailAST ast) {
+        return ast.getLastChild().getType() == TokenTypes.OBJBLOCK;
+    }
+
+    /**
+     * Check first array init matches given {@code name} and there are no more init expressions.
+     *
+     * @param declarator array declarator to check
+     * @param name expr ident to compare against
+     * @return {@code true} if first array init expression is {@link TokenTypes#IDENT} and
+     *     same as {@code name} and there are no more expressions (if array is multidimensional),
+     * {@code false} otherwise
+     */
+    private static boolean isArgUsedInArrayInit(DetailAST declarator, String name) {
+        final DetailAST nestedDecl = declarator.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
+        final DetailAST expr = declarator.findFirstToken(TokenTypes.EXPR);
+        boolean result = false;
```

Comment:
can it be `final boolean result;` ?

Reply:
It is false by default.
I can make it final and add else block with `result = false`, but in my opinion it is better to leave it like this.

Reply:
Agree, I was not attentive.